web3/middleware/gas_price.py

```diff
@@ -0,0 +1,18 @@
+from cytoolz.dicttoolz import (
+    assoc,
+)
+
+
+def gas_price_middleware(make_request, web3):
+    """
+    Includes a gas price using the gas price strategy
+    """
+    def middleware(method, params):
+        if method == 'eth_sendTransaction':
+            transaction = params[0]
+            generated_gas_price = web3.eth.generateGasPrice(transaction)
+            if 'gasPrice' not in transaction and generated_gas_price:
```

Comment:
I think that the check for `if 'gasPrice' not in transaction` should happen prior to the call to `web3.eth.generateGasPrice(...)` so that we can avoid the effort of computing the gas price if one was explicitly specified.